File: leetcode-04/solutions/cpp/0238.cpp
class Solution {
 public:
  vector<int> productExceptSelf(vector<int>& nums) {
    const int n = nums.size();

    vector<int> ans(n);
    ans[0] = 1;
    int r = 1;

    for (int i = 1; i < n; i++) ans[i] = nums[i - 1] * ans[i - 1];

    for (int i = n - 1; i >= 0; i--) {
      ans[i] *= r;
      r *= nums[i];
    }

    return ans;
  }
};

File: leetcode-04/solutions/python3/0238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)

        ans = [0] * n
        ans[0] = 1
        r = 1

        for i in range(1, n):
            ans[i] = nums[i - 1] * ans[i - 1]

        for i in range(n - 1, -1, -1):
            ans[i] *= r
            r *= nums[i]

        return ans
